### PR TITLE
[LoRA] Map inactive virtual-expert tokens to no-LoRA

### LINKED ISSUE #29157: [Bug] CUDA graph causes garbled outputs with concurrent multi-LoRA decoding on GLM-5.1-FP8
### Checklist

- [ ] I searched related issues but found no solution.
- [x] The bug persists in the latest version.
- [ ] Issues without environment info and a minimal reproducible demo are hard to resolve and may receive no feedback.
- [ ] If this is not a bug report but a general question, please start a discussion at https://github.com/sgl-project/sglang/discussions. Otherwise, it will be closed.
- [ ] Please use English. Otherwise, it will be closed.

### Describe the bug

# CUDA graph causes garbled outputs with concurrent multi-LoRA decoding on GLM-5.1-FP8

## Summary

We found a reproducible output corruption issue when serving GLM-5.1-FP8 with multiple LoRA adapters and CUDA graph enabled. Under concurrent decode, most requests collapse into repeated `!` tokens, while disabling CUDA graph makes the same workload clean.

This reproduces on clean upstream SGLang without any local patch/overlay.

## Launch Command

Example with 5 LoRAs and CUDA graph enabled:

```bash
export PYTHONPATH=/sgl-workspace/sglang/python

python3 -m sglang.launch_server \
  --model-path /path/to/GLM-5.1-FP8 \
  --served-model-name zai-org/glm-5.1 \
  --host 0.0.0.0 \
  --port 30000 \
  --tp 8 \
  --reasoning-parser glm45 \
  --tool-call-parser glm47 \
  --mem-fraction-static 0.93 \
  --enable-metrics \
  --enable-mfu-metrics \
  --disable-custom-all-reduce \
  --enable-lora \
  --lora-paths \
    r0_random=/path/to/random_loras/R0 \
    r1_random=/path/to/random_loras/R1 \
    r2_random=/path/to/random_loras/R2 \
    r3_random=/path/to/random_loras/R3 \
    r4_random=/path/to/random_loras/R4 \
  --max-loaded-loras 5 \
  --max-loras-per-batch 2 \
  --max-lora-rank 16 \
  --lora-use-virtual-experts \
  --disable-overlap-schedule \
  --cuda-graph-max-bs 16
```

Control run:

```bash
# Same command, but replace --cuda-graph-max-bs 16 with:
--disable-cuda-graph
```

## Reproduction Script

```python
import json
import urllib.request
import concurrent.futures

BASE = "http://127.0.0.1:30000"
PROMPT = "Write three sentences about the ocean."
MODELS = ["r0_random", "r1_random", "r2_random", "r3_random", "r4_random"]

N = 16
MAX_TOKENS = 128

def call(i):
    model = MODELS[i % len(MODELS)]
    body = json.dumps({
        "model": model,
        "prompt": PROMPT,
        "max_tokens": MAX_TOKENS,
        "temperature": 0
    }).encode()

    req = urllib.request.Request(
        BASE + "/v1/completions",
        data=body,
        headers={"Content-Type": "application/json"},
    )

    with urllib.request.urlopen(req, timeout=240) as r:
        d = json.loads(r.read())

    text = d["choices"][0]["text"]
    finish_reason = d["choices"][0].get("finish_reason")
    bad = ("!!!!!" in text) or (text.count("!") >= 20)

    return i, model, finish_reason, bad, text.count("!"), text[-100:]

with concurrent.futures.ThreadPoolExecutor(max_workers=N) as ex:
    rows = sorted(ex.map(call, range(N)))

for i, model, fin, bad, bangs, tail in rows:
    print(f"[req#{i:02d}] {model} fin={fin} "
          f"{GARBLED if bad else clean} bangs={bangs} tail={tail!r}")

print("SUMMARY garbled=%d total=%d" % (
    sum(1 for row in rows if row[3]), len(rows)
))
```

## Actual Result

With CUDA graph enabled, most concurrent requests collapse into repeated `!` tokens.

Example:

```text
[req#00] r0_random fin=length clean   bangs=0
[req#01] r1_random fin=length GARBLED bangs=70 tail=" the oceans density!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
[req#02] r2_random fin=length GARBLED bangs=70 tail= the oceans density!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!"
...
SUMMARY garbled=15 total=16
```

## Control Result

With `--disable-cuda-graph`, the same workload is clean:

```text
SUMMARY garbled=0 total=16
```

## Test Matrix

We tested randomly initialized LoRAs from 2 to 6 adapters. All use the same server settings except CUDA graph on/off.

| Random LoRA count | CUDA graph on | CUDA graph off |
|---:|---:|---:|
| 2 | 15/16 garbled | 0/16 garbled |
| 3 | 15/16 garbled | 0/16 garbled |
| 4 | 15/16 garbled | 0/16 garbled |
| 5 | 15/16 garbled | 0/16 garbled |
| 6 | 15/16 garbled | 0/16 garbled |

We also reproduced this with 5 trained LoRAs:

| LoRA type | CUDA graph on | CUDA graph off |
|---|---:|---:|
| 5 trained LoRAs | 15/16 garbled | 0/16 garbled |

## Notes

The random LoRAs were initialized with valid PEFT LoRA structure, rank 16, same target modules as the trained adapters. We used small random `lora_A` weights and zero `lora_B`, so the adapters are effectively near no-op. The corruption still reproduces, which suggests the issue is not caused by LoRA training quality or adapter content.

The key observation is that greedy decoding with `temperature=0` and identical prompts should be deterministic and consistent across requests. Instead, with CUDA graph enabled, most batch slots collapse into repeated `!` tokens. Disabling CUDA graph removes the issue.

This looks like a CUDA graph replay / static buffer / batch-slot state corruption issue in concurrent decode with multi-LoRA enabled, possibly related to GLM-5.1 FP8 + DSA decode backend (`fa3`) and LoRA CUDA graph buffers.

## Environment

- SGLang: 0.5.13
- Model: `zai-org/GLM-5.1-FP8`
- TP: 8
- GPUs: 8x GPU, each ~144GB memory
- KV cache dtype/backend observed in logs:
  - `Set DSA backends for bfloat16 KV Cache: prefill=flashmla_sparse, decode=fa3`
- LoRA:
  - `--enable-lora`
  - `--lora-use-virtual-experts`
  - `--max-lora-rank 16`
  - `--max-loras-per-batch 2`
  - tested with 2-6 LoRA adapters
- Clean SGLang source only:
  - `PYTHONPATH=/sgl-workspace/sglang/python`
  - no local monkey patch / overlay loaded

### PR DESCRIPTION
**Pull request by ronhuafeng**
## Motivation

Fixes #29157.

MoE virtual-expert LoRA uses `token_lora_mapping == -1` as the no-LoRA sentinel, but `_compute_moe_lora_info` could map rank-0 or inactive request segments to adapter slot `0`. In CUDA graph replay, the preallocated mapping buffer can also keep stale values outside the active token prefix. That lets inactive or padded tokens look like they should route through LoRA slot 0, which can corrupt decode output under concurrent multi-LoRA serving.

This PR makes the producer side enforce the invariant directly:

```text
token_lora_mapping[token] >= 0  <=>  valid adapter id with rank > 0
otherwise token_lora_mapping[token] == -1
```

## Modifications

- Map rank-0 LoRA segments to `-1` in both the Triton kernel path and the torch fallback path.
- Guard negative request-to-LoRA ids before reading `lora_ranks`, preserving `-1` as no-LoRA instead of indexing slot 0 or the tail.
- Reset the full preallocated CUDA graph `token_lora_mapping` buffer to `-1` before slicing the active token prefix.
- Extend `test_moe_lora_info.py` to cover rank-0 segments, negative no-LoRA segments, and stale preallocated-buffer tails.
- Fold in automated review feedback to avoid redundant PyTorch indexing and temporary `zeros_like` / `full_like` allocations in the fallback/test helper path.

## Accuracy Tests

End-to-end serving validation was run on an 8x H200-class NVIDIA setup with GLM-5.1-FP8, TP=8, `--enable-lora`, `--lora-use-virtual-experts`, 5 LoRA adapters, and concurrent `/v1/completions` requests:

| case | CUDA graph | result |
| --- | --- | --- |
| baseline | enabled | 15/16 garbled |
| control | disabled | 0/16 garbled |
| patched | enabled | 0/16 garbled |

The final PR also adds a small guard for pre-existing `-1` request mappings.

Validation for current head `d9c09011ee198ae4127679def9faff2827d33169`:

- Remote `Lint`: passed, including pre-commit checks, docs offline link checks, and `sgl-kernel` clang-format checks.
- Local checks:

```bash
python3 -m py_compile python/sglang/srt/lora/backend/base_backend.py test/registered/lora/test_moe_lora_info.py
git diff --check
```

Focused CUDA unit test not run locally:

```bash
python3 -m pytest test/registered/lora/test_moe_lora_info.py -q
```

The local machine used to prepare this PR does not have `pytest`, `torch`, `triton`, or CUDA test dependencies installed.

## Speed Tests and Profiling

No dedicated speed benchmark was run. The runtime change is limited to LoRA metadata preparation and CUDA graph buffer reset for MoE LoRA mapping.

## Checklist

- [x] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [x] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/benchmark_and_profiling.html).
- [x] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

---
### CI States

Latest PR Test (Base): :x: [Run #28273836930](https://github.com/sgl-project/sglang/actions/runs/28273836930)
Latest PR Test (Extra): :x: [Run #28273836864](https://github.com/sgl-project/sglang/actions/runs/28273836864)

### CODE REVIEW

**Comment by ronhuafeng** [comment]
Ready for review. I addressed the automated review suggestions in the current single commit `d9c09011ee198ae4127679def9faff2827d33169`, and the remote Lint workflow passes on this head. Local checks also passed for `py_compile` and `git diff --check`; the focused CUDA pytest still needs a matching test environment or maintainer CI.

The PR is still missing the `run-ci` label, so full PR tests are gated. Could a maintainer please trigger CI for the current commit with `/tag-and-rerun-ci`?

**Comment by ronhuafeng** [comment]
Current-head validation update for `d9c09011ee198ae4127679def9faff2827d33169`:

Reran the end-to-end repro on a GCP Spot `a3-ultragpu-8g-nolssd` VM in `team-huafeng/europe-west1-b` with 8x NVIDIA H200, `lmsysorg/sglang:v0.5.13-cu129`, `zai-org/GLM-5.1-FP8`, TP=8, `--enable-lora`, `--lora-use-virtual-experts`, 5 shared-outer random MoE LoRAs, and 16 concurrent `/v1/completions` requests.

Results:

| case | CUDA graph | result |
| --- | --- | --- |
| baseline | enabled, `--cuda-graph-max-bs 16` | `15/16` garbled |
| control | disabled, `--disable-cuda-graph` | `0/16` garbled |
| patched current PR head | enabled, `--cuda-graph-max-bs 16` | `0/16` garbled |

Local sanity checks on the clean PR worktree also passed:

```bash
git diff --check origin/main...HEAD
python3 -m py_compile python/sglang/srt/lora/backend/base_backend.py test/registered/lora/test_moe_lora_info.py
```

The Spot VM has been deleted after the run; the cache/data disk was preserved. Full PR tests still appear gated on `pr-gate` / missing maintainer CI trigger.

File: python/sglang/srt/lora/backend/base_backend.py [hunk 1]
```diff
@@ -420,9 +429,20 @@ def _compute_moe_lora_info(
...
+                0, valid_indices.long(), (active_ranks > 0).to(torch.int32)
+            )
+        safe_weight_indices = torch.where(
+            valid_weight_indices, weight_indices, torch.zeros_like(weight_indices)
+        )
+        active_ranks = torch.where(
+            valid_weight_indices,
+            lora_ranks[safe_weight_indices.long()],
+            torch.zeros_like(weight_indices),
         )
```
[the remark this answers is not in the document]
Comment (ronhuafeng) [reply python/sglang/srt/lora/backend/base_backend.py:438]: Thanks. This is reflected in the current head: `active_ranks` is computed once using safe indices, and `adapter_enabled` is populated from `active_ranks[valid_weight_indices]`, so we avoid the duplicate rank lookup here.

File: python/sglang/srt/lora/backend/base_backend.py [hunk 2]
```diff
@@ -440,8 +460,13 @@ def _compute_moe_lora_info(
         torch.searchsorted(seg_indptr.to(torch.int32), token_positions, right=True) - 1
     )
 
+    active_weight_indices = torch.where(
+        valid_weight_indices & (active_ranks > 0),
+        weight_indices.to(torch.int32),
+        torch.full_like(weight_indices, -1, dtype=torch.int32),
+    )
```
[the remark this answers is not in the document]
Comment (ronhuafeng) [reply python/sglang/srt/lora/backend/base_backend.py:466]: Thanks. The current head already uses the scalar `-1` form in `torch.where`, so this no longer allocates a `full_like` tensor for the fallback value.

File: test/registered/lora/test_moe_lora_info.py [hunk 3]
```diff
@@ -14,14 +14,38 @@ def _expected_adapter_enabled(
...
+    active_ranks = torch.where(
+        valid_weight_indices,
+        lora_ranks[safe_weight_indices.long()],
+        torch.zeros_like(weight_indices),
+    )
+    active_weight_indices = torch.where(
+        valid_weight_indices & (active_ranks > 0),
+        weight_indices,
+        torch.full_like(weight_indices, -1),
+    )
```
[the remark this answers is not in the document]
Comment (ronhuafeng) [reply test/registered/lora/test_moe_lora_info.py]: This thread is outdated against the current diff. The production path now uses the scalar `0` / `-1` form; I do not think the outdated test-only suggestion needs a separate follow-up change.

**Comment by ronhuafeng** [comment]
I ran a baseline-vs-patched runtime check for this PR.

Setup:
- Hardware: 8x NVIDIA H200.
- Image: `lmsysorg/sglang:v0.5.13-cu129@sha256:31f10bd08c4d7691a31add7406a3016479b606d51af1be1479ad7756866a840b`.
- Model: `zai-org/GLM-5.1-FP8`, served as `zai-org/glm-5.1`.
- PR head tested: `d9c09011ee198ae4127679def9faff2827d33169`.
- LoRA setup: 5 synthetic shared-outer MoE LoRAs, rank 16, bf16, first MoE layer 3, hidden size 6144, MoE intermediate size 2048, `--enable-lora`, `--lora-use-virtual-experts`, `--max-loaded-loras 5`, `--max-loras-per-batch 2`.
- Probe: 16 concurrent OpenAI completions across the 5 adapters, `max_tokens=128`.

Results:
- Baseline with CUDA graph enabled reproduced the regression: 15 / 16 responses were classified as garbled.
- Baseline with CUDA graph disabled was clean: 0 / 16 garbled.
- Patched with CUDA graph enabled was clean: 0 / 16 garbled.

So this patch fixed the CUDA-graph-specific virtual-expert LoRA corruption in the focused GLM-5.1 FP8 reproduction I ran.

<details>
<summary>Repro shape</summary>

Baseline and patched runs used the same container image, model snapshot, generated LoRA set, prompt shape, and 16-request probe. The only behavioral comparison dimensions were:

1. baseline, CUDA graph enabled, no PR overlay, expected garbled output;
2. baseline, CUDA graph disabled, no PR overlay, expected clean output;
3. patched, CUDA graph enabled, PR head overlay, expected clean output.

The probe counted output as garbled when generated text contained repeated exclamation-mark bursts or similar obvious corruption, and clean otherwise.
</details>
